docs/adr/adr-008-monitoring-and-alerting.md

```diff
@@ -0,0 +1,277 @@
+[MetricsCollectionWorker][] are succeeding at least once within a time frame
+(suggestion is 1 hour). These jobs should continue running even when the
+system is busy and the lack of them running indicates that we no longer have
+confidence the system is monitored or able to recover. Thus, an alert will be
+created that contacts engineers out-of-hours should these not succeed in a
+time frame.
+
+For Sidekiq workers that don't affect the user noticeable aspects of Email
+Alert API, such as [EmailDeletionWorker][], we won't raise critical alerts if
+these are failing. Instead we will raise a warning if any Sidekiq workers
+exhaust all their retries and are placed in the [dead set][sidekiq-dead]. This
+approach was chosen as it was a pragmatic technique to monitor multiple workers
+without adding alerts for individual workers. Workers will have their
+[retry cadence][sidekiq-retry] considered so that this will alert in a
+reasonable time frame of the retries being exhausted (the Sidekiq default of 21
+days is very long).
+
+We will modify the [latency checks for email queues][latency-checks]. We
+will be treating [transactional emails][] different to other latency checks
+as these emails have have an [expectation of a prompt
+delivery][transactional-delivery-expectation] and the lack of them blocks a
+user from progressing with their task. These will raise a warning after 5 mins
+and go critical (with engineers contacted out-of-hours) should the latency
+reach 15 mins.
+
+The other latency checks will be made significantly less sensitive to latency
+and we will change the time before a warning occurs to be in the measure of
+multiple hours compared to minutes, with the intention that they serve to
+alert a support engineer only when the system is exceptionally busy and may be
+unhealthy. These alerts will only reach a critical status if they hit a 24 hour
+point as this would indicate that we are consistently creating email faster
+than the application can send it, which may mean the system is perpetually
+backlogged.
+
+#### Web application
+
+We decided that there should be a mechanism that alerts people out-of-hours if
+the Email Alert API web application appears broken, noting that currently an
+engineer is unlikely to be contacted in the event of the whole Email Alert API
+application becoming unavailable - unless travel advice were published. We
+noted there wasn't a particularly consistent approach to this across GOV.UK
+([with some apps raising an out-of-hours alert when a single instance fails
+healthchecks][publishing-api-alert]) and felt it would be good to take a simple
+approach to avoid surprising engineers.
+
+We decided the most appropriate place to alert would be when [the application
+load balancer has no healthy instances of the application][aws-lb-check]. This
+represents that no instances of the web application are healthy and
+[no requests will succeed][aws-alb-health].
```

Comment:
nit: all the load balancer healthchecks failing doesn't _necessarily_ mean that user requests aren't succeeding, because in that situation the load balancer will behave as if all backends are healthy (because the healthchecks might be returning false negatives so it might as well try — and this type of scenario isn't quite as rare as we might be tempted to imagine).

I think it's still a reasonable surrogate measure (if a little unorthodox) and probably a good tradeoff for now in terms of implementation simplicity. It's also no bad thing to alert in the case where the healthchecks are all broken, even if the service might still be up - since that scenario would mean the alerting is blind to any real outage that might happen until the healthchecks are fixed.

Reply:
The reason I mention this is that it's something worth mentioning when someone comes to write the playbook entry for this alert — i.e. there are two possible causes: either the backends are all unavailable or (less likely but by no means impossible) the healthchecks are broken or timing out and therefore the alert is broken but the service may or may not still be up.

Reply:
@sengi Thanks for your comments ⭐ 

The behaviour for no healthy instances varies whether you are using an [application load balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html) or a [classic load balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/introduction.html). Key phrasing:

Application Load Balancer: "If a target group contains only unhealthy registered targets, the load balancer nodes route requests across its unhealthy targets."
Classic Load Balancer:  "The load balancer routes requests only to the healthy instances. When the load balancer determines that an instance is unhealthy, it stops routing requests to that instance"

As Email Alert API uses a classic load balancer it will serve 503 responses if all instances of the app are failing the healthcheck - we found this out when the whole app went down and we were wondering where all the stats were (we anticipated high 5xx rate for instance).

Email Alert Frontend is behind an application load balancer so will switch to serving to all instances if all are unhealthy.

---

The tricky bit around this is trying to establish if there is a single point that can be monitored for web application health. If just one box is broke (which admittedly doesn't seem to have much precedent) it doesn't feel necessary to call some up for one, nor is it much fun for a support engineer to get different calls for every box if they all break together. This all seems to tie into smells that are alerting system is a bit outdated as I imagine there are better approaches to all this that the rest of the world have adopted.

Reply:
Oh gosh that's shocking! You're absolutely right though. I hadn't come across that before but it's definitely the case — classic ELB responds `503 Service Unavailable: Back-end server is at capacity` when all its backends are unhealthy.

If I'd known this I would have replaced these with ALBs during the AWS migration 😞

We might want to increase the healthcheck timeout in light of this. It's only 3 seconds at the moment whereas I'm sure the timeout on the app itself is longer than that (the nginx timeout is 15 s, couldn't easily find where the Rails one is set).

Reply:
Yeah, it was a surprise for me too. To be honest I find it makes quite a lot of sense logically that it's genuinely severe when nothing is healthy - whereas you seem to enter this frustratingly ambiguous world when all healthchecks fail means you're serving all instances (compared to say the oddness of 15 of 16 instances failing healthchecks and you only serving from).

I'm assuming though that "classic" is a synonym for legacy so this behaviour is on it's way out.

There tends to not be a Rails timeout, I think when that is done it is implemented in Rack and I've not seen that we've done that anywhere (we also have plenty of evidence of Rails requests cruising to take > 30s). 15s would sound a better number - I have a feeling that icinga may time out at 10s for healthchecks so could be consistent with that. Mostly the healthchecks should be very fast, but there is a chance they could get stalled if one of their dependent services (a db for instance) has some sort of connection issue.

Reply:
Yeah we actually saw that kind of timeout scenario for real in integration while working on the AWS migration (must have been a different app though because we didn't find out about the surprising blackhole behaviour of the load balancer!) Classic ELB is the first load balancer I've come across which doesn't have some sort of last-resort mode (it's a common enough concept that it has a name). Oh well, the more you know!